src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
+                    let next = reach.perform_access2(step);
+                    match &next {
+                        Ok(configuration) =>
+                            assert!(
+                                seen.contains(&next),
+                                "{reach} is reachable from {state}, but the extra step {step:?} leads to {configuration} which is not registered as reachable"
+                            ),
+                        _ =>
                             assert!(
-                                kind1 == AccessKind::Read && kind2 == AccessKind::Read,
-                                "failed to enforce noalias between two accesses that are not both reads"
-                            );
+                                seen.contains(&next),
+                                "UB can be triggered from {reach} by {step:?}, should be recorded as reachable from {state}"
+                            ),
+                    }
+                }
+            }
+        }
+    }
+
+    impl LocStateProt2 {
+        fn perform_opaque2(mut self, code: &Opaque2) -> Result<Self, ()> {
+            for evt in code.events.iter().flatten() {
+                self = self.perform_access2(*evt)?;
+            }
+            Ok(self)
+        }
+    }
+
+    impl fmt::Display for Opaque2 {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            for evt in self.events.iter().flatten() {
+                let kind = match evt.kind {
+                    AccessKind::Read => "read",
+                    AccessKind::Write => "write",
+                };
+                use AccessRel2 as Rel;
+                let ptr = match evt.rel2 {
+                    Rel::ChildX => "x",
+                    Rel::ChildY => "y",
+                    // conventional name for a common ancestor of `x` and `y` is chosen to be `z`.
+                    Rel::Other => "z",
+                };
+                write!(f, "{kind} {ptr}; ")?;
+            }
+            Ok(())
+        }
+    }
+
+    impl LocStateProt2 {
+        fn trivially_indistinguishable(&self, other: &Self) -> bool {
+            self == other
+        }
+
+        #[rustfmt::skip]
+        /// Two states (by convention `self` is the source and `other` is the target)
+        /// are "distinguishable" if there exists a sequence of instructions
+        /// that causes UB in the target but not in the source.
+        /// This (fast enough) implementation simply explores the reachable space
+        /// by all sequences of `Access2` or function returns.
+        fn distinguishable(&self, other: &Self) -> bool {
+            if self.trivially_indistinguishable(other) { return false; }
+            let mut states = vec![(self.clone(), other.clone())];
+            let mut seen = FxHashSet::default();
+            while let Some(state) = states.pop() {
+                if seen.contains(&state) { continue; }
+                seen.insert(state.clone());
```

Comment:
`seen.insert` returns whether the thing was in the set, so you don't need to do `contains` and then `insert` -- you can do this in a single step.